internal/helm/repository/chart_repository.go

```diff
@@ -308,7 +340,46 @@ func (r *ChartRepository) LoadFromFile(path string) error {
 	if err != nil {
 		return err
 	}
-	return r.LoadIndexFromBytes(b)
+
+	h := sha256.New()
+	if _, err := io.Copy(h, bytes.NewReader(b)); err != nil {
+		return err
+	}
+
+	checksum := fmt.Sprintf("%x", h.Sum(nil))
+	// If optimization is disabled, use YAML index loader and nothing else.
+	if !r.OptimizeIndexLoading {
+		return r.LoadIndex(b, checksum, indexFromYAML)
+	}
+
+	// First check that an optimized index exists, if so load using JSON index loader.
+	// On top of a more efficient unmarshaling process, it will also avoid a sorting
+	// operation across the entire index.
+	jsonIndex := filepath.Join(filepath.Dir(path), fmt.Sprintf("index-%s.json", checksum))
+	if f, err := os.Lstat(jsonIndex); err == nil && f.Mode()&os.ModeSymlink == 0 {
+		if f.Size() > helm.MaxIndexSize {
+			return fmt.Errorf("size of optimized index '%s' exceeds '%d' bytes limit", stat.Name(), helm.MaxIndexSize)
+		}
+
+		cb, err := os.ReadFile(jsonIndex)
+		if err != nil {
+			return err
+		}
+		return r.LoadIndex(cb, checksum, indexFromJSON)
+	}
+
+	// Fallback to the YAML index loader.
+	if err = r.LoadIndex(b, checksum, indexFromYAML); err != nil {
+		return err
+	}
+
+	// Marshal the loaded index at this point, as it is fully loaded and sorted.
+	jb, err := json.Marshal(r.Index)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(jsonIndex, jb, 0o600)
```

Comment:
Our GC process needs to be modified to ignore this json file, currently all files present in the artifact directory except the actual artifact (`index.yaml` in this case) are deleted.